CenterForOpenScience/osf.io
# [feature] bapi 502s on deleted files [OSF-8963]

## Purpose

Fix issues where browsable API 502s on deleted files.

For files, we are determining the serializer_class dynamically (can be either `FileDetailSerializer` or `QuickFilesDetailSerializer`) by looking up the _id or guid from the URL. Sometimes, the file has already been deleted in which case, NotFound (_id) or Gone(guid) is raised. With the current version of DRF, it seems, any exceptions raised in `get_serializer_class` results in 502s. One of the solutions; we should not raise any errors when `get_file` is called from the serializer_class in case the object is found (does not matter if `is_deleted` is `True`. In case NotFound is raised just default to FileDetailSerializer. https://github.com/CenterForOpenScience/osf.io/blob/2881a189684f4e1389fe30633b59a36ee2f31695/api/files/views.py#L37-L50

## Changes
Add `ser` bool flag to tell when get_file is called from `get_serializer_class` or `get_object`. in the first case, we only expect an object and we handle NotFound by defaulting to `FileDetailSerializer`.

## QA Notes
- This fix pertains to the browsable API, hence API testing is required.
- Make sure for deleted file: url of the sorts https://staging-api.osf.io/v2/files/4r523/ and https://staging-api.osf.io/v2/files/5a56835b08a8d80206c787e5/ do not 502.
- Try the same urls in postman or any other similar API testing tool.

## Side Effects

NA

## Ticket
[OSF-8963](https://openscience.atlassian.net/browse/OSF-8963)
